File: api/src/test/java/net/blueshell/api/factory/model/BoardMemberFactory.java
package net.blueshell.api.factory.model;

import net.blueshell.api.testutil.ModelTestUtils;
import com.github.javafaker.Faker;
import lombok.RequiredArgsConstructor;
import net.blueshell.api.model.File;
import net.blueshell.api.model.User;
import net.blueshell.api.model.board.Board;
import net.blueshell.api.model.board.BoardMember;
import org.springframework.stereotype.Component;

import java.util.concurrent.atomic.AtomicLong;

/**
 * Factory for BoardMember model test instances.
 */
@Component
@RequiredArgsConstructor
public class BoardMemberFactory {

    private static final AtomicLong counter = new AtomicLong(1000);
    private final Faker faker;
    private final BoardFactory boardFactory;
    private final UserFactory userFactory;
    private final FileFactory fileFactory;

    public BoardMember createBasic(
            Board board,
            User user
    ) {
        BoardMember bm = new BoardMember();

        File picture = fileFactory.createImage();

        bm.setBoard(board);
        bm.setUser(user);
        bm.setPicture(picture);

        return bm;
    }

    public BoardMember createFull(Board board, User user) {
        return createBasic(board, user);
    }

    public BoardMember createWithCustomizations(Board board, User user, java.util.function.Consumer<BoardMember> customizer) {
        BoardMember bm = createFull(board, user);
        customizer.accept(bm);
        return bm;
    }

    private Long generateId() {
        return counter.incrementAndGet();
    }
}

File: api/src/test/kotlin/net/blueshell/api/factory/model/BoardMemberFactory.kt
package net.blueshell.api.factory.model

import com.github.javafaker.Faker
import net.blueshell.api.model.User
import net.blueshell.api.model.board.Board
import net.blueshell.api.model.board.BoardMember
import org.springframework.stereotype.Component
import java.util.concurrent.atomic.AtomicLong
import java.util.function.Consumer

/**
 * Factory for BoardMember model test instances.
 */
@Component
class BoardMemberFactory(
    private val faker: Faker,
    private val boardFactory: BoardFactory,
    private val userFactory: UserFactory,
    private val fileFactory: FileFactory
) {

    fun createBasic(board: Board, user: User): BoardMember {
        val boardMember = BoardMember()
        val picture = fileFactory.createImage()
        boardMember.board = board
        boardMember.user = user
        boardMember.picture = picture
        return boardMember
    }

    fun createFull(board: Board, user: User): BoardMember = createBasic(board, user)

    fun createWithCustomizations(board: Board, user: User, customizer: Consumer<BoardMember>): BoardMember {
        val boardMember = createFull(board, user)
        customizer.accept(boardMember)
        return boardMember
    }

    private fun generateId(): Long = COUNTER.incrementAndGet()

    private companion object {
        val COUNTER = AtomicLong(1000)
    }
}
